### PR TITLE
SOLR-17120: handle null value when merging partials

### PR DESCRIPTION
**Pull request by eukaryote**
https://issues.apache.org/jira/browse/SOLR-17120

Avoid a `NullPointerException` that can occur under some circumstances when performing multiple partial updates of the same document in separate requests using `commitWithin`.

# Description

When merging older partial updates into the doc for newer partial updates, the existing solr code would generate a `NullPointerException` under some circumstances when it tried to iterate over a `null` value.

# Solution

I replaced the loop that started with `for (Object val : olderDoc.getFieldValues(fieldName))` with code that first assigns the value to a temporary variable and then either iterates over the value (performing the same action as  previously on each value if it is non-nulll) or calls `newerDoc.addField(fieldName, null)` instead if the value is `null`.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

No tests yet, but I was able to regularly reproduce the error before, and have been running a patched version of a `solr-core-9.4.0-SNAPSHOT.jar` file that I created with this change against the 9.4.0 version and haven't been able to reproduce the error.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://github.com/apache/solr/blob/main/CONTRIBUTING.md) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [x] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [ ] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

21 tests failed, all of which were in `org.apache.solr.s3.S3BackupRepositoryTest` or `org.apache.solr.s3.S3OutputStreamTest`, which my change doesn't seem related to at all. EDIT: After I removed my aws configs and ran the tests again, they all passed.

### CODE REVIEW

File: solr/CHANGES.txt [hunk 1]
```diff
@@ -164,6 +164,9 @@ Bug Fixes
 
 * SOLR-17090: The v2 "delete alias" API no longer errantly returns a 405 status code (Jason Gerlowski)
 
+* SOLR-17120: Fix NullPointerException in UpdateLog.applyOlderUpdates that can occur if there are multiple partial
+  updates of the same document in separate requests using commitWithin. (Calvin Smith, Christine Poerschke)
+
```
Comment (cpoerschke) [inline solr/CHANGES.txt:169]: Tentatively adding for 9.5.0 here but we can move it to 9.6.0 if it's too late for the 9.5.0 release.

**Review by cpoerschke — APPROVED** [review]

**Comment by gerlowskija** [comment]
> 21 tests failed, all of which were in org.apache.solr.s3.S3BackupRepositoryTest or org.apache.solr.s3.S3OutputStreamTest

Solr has some test-flakiness issues, and the S3 tests are common culprits.  Getting tests to run reliably is a bit of a dark-art.  We try to get around this as best we can by fixing flakiness when we find it, running multiple iterations, etc.

TL;DR - I wouldn't worry about these failures too much; whoever ends up merging this can validate that tests run for them 👍

File: solr/CHANGES.txt (hunk 1 above)
Comment (gerlowskija) [reply solr/CHANGES.txt]: +1 to target 9.5.0 - seems like a pretty safe change!

**Review by gerlowskija — APPROVED** [review]
LGTM - do you want to handle merging @cpoerschke , or should I?  Don't want to step on your toes here...
